src/appleseed/foundation/image/genericimagefilewriter.cpp

```diff
@@ -31,48 +30,488 @@
+{
+    assert(image);
+
+    impl->m_canvas.push_back(image);
+    impl->m_spec.push_back(OIIO::ImageSpec{});
+
+    set_image_spec();
+}
+
+namespace
+{
+    OIIO::TypeDesc convert_pixel_format(const PixelFormat format)
+    {
+        switch (format)
+        {
+          case PixelFormatUInt8:
+              return OIIO::TypeDesc::UINT8;
+
+          case PixelFormatUInt16:
+              return OIIO::TypeDesc::UINT16;
+
+          case PixelFormatUInt32:
+              return OIIO::TypeDesc::UINT32;
+
+          case PixelFormatHalf:
+              return OIIO::TypeDesc::HALF;
+
+          case PixelFormatFloat:
+              return OIIO::TypeDesc::FLOAT;
+
+          case PixelFormatDouble:
+              return OIIO::TypeDesc::DOUBLE;
+
+          default:
+              return OIIO::TypeDesc::UNKNOWN;
+        }
+    }
+}
+
+void GenericImageFileWriter::set_image_output_format(const PixelFormat output_pixel_format)
+{
+    assert(!impl->m_spec.empty());
+
+    impl->m_spec.back().set_format(convert_pixel_format(output_pixel_format));
+}
+
+void GenericImageFileWriter::set_image_channels(
+    const size_t    channel_count,
+    const char**    channel_names)
+{
+    assert(!impl->m_spec.empty());
+    assert(channel_count > 0);
+    assert(channel_names);
+
+    OIIO::ImageSpec& spec = impl->m_spec.back();
+
+    spec.nchannels = static_cast<int>(channel_count);
+
+    for (size_t i = 0; i < channel_count; i++)
+    {
+        const char* name = channel_names[i];
+
+        spec.channelnames.push_back(name);
+
+        if (strcmp(name, "A") == 0)
+            spec.alpha_channel = static_cast<int>(i);
+    }
+}
+
+void GenericImageFileWriter::set_image_spec()
+{
+    assert(!impl->m_spec.empty());
+
+    const CanvasProperties& props = impl->m_canvas.back()->properties();
+    OIIO::ImageSpec& spec = impl->m_spec.back();
+
+    // Size of the data of the image.
+    spec.width = static_cast<int>(props.m_canvas_width);
+    spec.height = static_cast<int>(props.m_canvas_height);
+
+    // Origin of the pixel data of the image.
+    spec.x = 0;
+    spec.y = 0;
+
+    // Full size of the data of the image.
+    spec.full_width = spec.width;
+    spec.full_height = spec.height;
+
+    // Origin of the pixel data of the full image.
+    spec.full_x = spec.x;
+    spec.full_y = spec.y;
+
+    // Size of a tile.
+    if (m_writer->supports("tiles"))
+    {
+        spec.tile_width = static_cast<int>(props.m_tile_width);
+        spec.tile_height = static_cast<int>(props.m_tile_height);
+    }
+    else
+    {
+        spec.tile_width = 0;
+        spec.tile_height = 0;
+    }
+
+    // Channel names.
+    const char* channel_names[] = { "R", "G", "B", "A" };
+    set_image_channels(props.m_channel_count, channel_names);
+
+    // Format of the pixel data.
+    const boost::filesystem::path filepath(m_filename);
+    const std::string extension = lower_case(filepath.extension().string());
+
+    if (extension == ".png")
```

Comment:
What's special about PNG here? Why are we making a special case for PNG and not, say, BMP (or any of the 8-bit per channel file format)?

Reply:
Because we don't support BMP file format.
Theses lines set the default output channel format for each of the supported file formats.
OIIO make same (PNG's default output channel format is also 8-bit).

We can delete theses lines. But we'll need to set the output channel format manually wherever we write file on disk with the ```GenericImageFileWriter``` or leave OIIO choose for us.

Reply:
What would happen if we always did

    set_image_output_format(props.m_pixel_format);

Reply:
For formats that don't supports that output pixel format, it takes the OIIO's default one

Reply:
We must to understand that the raw pixel image buffer must be in the same type than ```props.m_pixel_format```.

OIIO will convert that row pixel image buffer in its desired type.

If it's not the case. That crash